src/NuGet.Indexing/ExpandAcronymsFilter.cs

```diff
@@ -0,0 +1,118 @@
+
+        private readonly ITermAttribute _termAttribute;
+        private readonly IPositionIncrementAttribute _positionIncrementAttribute;
+        private readonly HashSet<string> _tokenSet;
+        private State _currentState;
+
+        public ExpandAcronymsFilter(TokenStream input, IAcronymExpansionProvider acronymExpansionProvider) 
+            : base(input)
+        {
+            _acronymExpansionProvider = acronymExpansionProvider;
+
+            _termAttribute = AddAttribute<ITermAttribute>();
+            _positionIncrementAttribute = AddAttribute<IPositionIncrementAttribute>();
+            _tokenSet = new HashSet<string>();
+        }
+
+        public override bool IncrementToken()
+        {
+            if (_tokenSet.Count > 0)
+            {
+                RestoreState(_currentState);
+
+                var tokenAtHead = _tokenSet.First();
+
+                _termAttribute.SetTermBuffer(tokenAtHead);
+                _tokenSet.Remove(tokenAtHead);
+
+                _positionIncrementAttribute.PositionIncrement = 0;
+
+                return true;
+            }
+
+            try
+            {
+                if (!input.IncrementToken()) // end of stream; no more tokens on input stream
+                {
+                    return false;
+                }
+            }
+            catch (Exception ex)
+            {
+                return false;
+            }
+
+            if (!string.IsNullOrEmpty(_termAttribute.Term))
+            {
+                var acronyms = _acronymExpansionProvider.GetKnownAcronyms()
+                    .Where(a => _termAttribute.Term.IndexOf(a, StringComparison.OrdinalIgnoreCase) >= 0);
+
+                foreach (var acronym in acronyms)
+                {
+                    // Add expanded acronym (ef => entity;framework)
+                    foreach (var expansion in _acronymExpansionProvider.Expand(acronym))
+                    {
+                        _tokenSet.Add(expansion);
+                    }
+
+                    // Add original term without the acronym (xamlbehaviors with xaml acronym => behaviors)
+                    var termWithoutAcronym = RemoveSubstring(_termAttribute.Term, acronym);
+                    if (!string.IsNullOrEmpty(termWithoutAcronym))
+                    {
+                        _tokenSet.Add(termWithoutAcronym);
+                    }
+                }
+            }
+
+            _currentState = CaptureState();
+            return true;
+        }
+
+        private static string RemoveSubstring(string original, string substring)
+        {
+            if (string.IsNullOrEmpty(original) || string.IsNullOrEmpty(substring))
+            {
+                return original;
+            }
+
+            var result = new StringBuilder(original.Length);
+
+            int substringLength = substring.Length;
+            int substringStartIndex = -1;
+            int lastCharacterIndex = 0;
+            
+            while (true)
```

Comment:
`while (true) ... if (condition == true) break;` is perhaps more clear when rewritten to use `while (condition == false)`?
e.g. `while (substringStartIndex >= 0)`